axum/src/extract/host.rs

```diff
@@ -0,0 +1,59 @@
+use super::{FromRequest, RequestParts};
+use async_trait::async_trait;
+use std::{convert::Infallible};
+
+/// Extractor that extracts the host from a request.
+#[derive(Debug, Clone, Default)]
+pub struct Host(pub String);
+
+#[async_trait]
+impl<B> FromRequest<B> for Host
+where
+    B: Send,
+{
+    type Rejection = Infallible;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        if let Some(host) = req.uri().host() {
+            return Ok(Host(host.to_string()));
+        }
+
+        if let Some(Ok(host)) = req.headers().get("host").map(|host| host.to_str()) {
+            return Ok(Host(host.to_string()));
+        }
+
+        Ok(Host("".to_string()))
```

Comment:
Hm I'm not sure what is right here. Returning an empty host feels odd 🤔 Maybe we should return a rejection instead. Then the user can use `Option<Host>` to explicitly make hosts optional.

Reply:
I like this, I initially wanted to return an Option but couldn't quite work out the types. I didn't realise the result could be converted into an option if that's how it was used in the fn def. What makes that work?

Reply:
I've created an error type for this method, please check :)

Reply:
Looks good!

The `Option` and `Result` extract trick works because of https://github.com/tokio-rs/axum/blob/main/axum-core/src/extract/mod.rs#L217-L241